executor/infoschema_reader.go

```diff
@@ -1426,6 +1427,125 @@ func (e *memtableRetriever) setDataFromTableConstraints(ctx sessionctx.Context,
 	e.rows = rows
 }
 
+// tableStorageStatsRetriever is used to read slow log data.
+type tableStorageStatsRetriever struct {
+	dummyCloser
+	table         *model.TableInfo
+	outputCols    []*model.ColumnInfo
+	retrieved     bool
+	initialized   bool
+	extractor     *plannercore.TableStorageStatsExtractor
+	initialTables []*initialTable
+	curTable      int
+	pdAddress     string
+	helper        *helper.Helper
+	stats         helper.PDRegionStats
+}
+
+func (e *tableStorageStatsRetriever) retrieve(ctx context.Context, sctx sessionctx.Context) ([][]types.Datum, error) {
+	if e.retrieved {
+		return nil, nil
+	}
+	if !e.initialized {
+		err := e.initialize(sctx)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if len(e.initialTables) == 0 || e.curTable >= len(e.initialTables) {
+		e.retrieved = true
+		return nil, nil
+	}
+
+	rows, err := e.setDataForTableStorageStats(sctx)
+	if err != nil {
+		return nil, err
+	}
+	if len(e.outputCols) == len(e.table.Columns) {
+		return rows, nil
+	}
+	retRows := make([][]types.Datum, len(rows))
+	for i, fullRow := range rows {
+		row := make([]types.Datum, len(e.outputCols))
+		for j, col := range e.outputCols {
+			row[j] = fullRow[col.Offset]
+		}
+		retRows[i] = row
+	}
+	return retRows, nil
+}
+
+type initialTable struct {
+	db string
+	*model.TableInfo
+}
+
+func (e *tableStorageStatsRetriever) initialize(sctx sessionctx.Context) error {
+	is := infoschema.GetInfoSchema(sctx)
+	dbs := is.AllSchemas()
+	sort.Sort(infoschema.SchemasSorter(dbs))
+	schemas := e.extractor.TableSchema
+	tables := e.extractor.TableName
+
+	// Extract the tables to the initialTable.
+	for _, schema := range dbs {
+		if len(schemas) == 0 || schemas.Exist(schema.Name.L) {
+			for _, table := range schema.Tables {
+				if len(tables) == 0 || tables.Exist(table.Name.L) {
+					e.initialTables = append(e.initialTables, &initialTable{schema.Name.L, table})
+				}
+			}
+		}
+	}
+	if len(e.initialTables) == 0 {
+		return errors.Errorf("schema or table not exist, please check the schema and table")
```

Comment:
I think the error is redundant,  we can remove it.

Reply:
OK, removed it.